### DISCUSSION TITLE
Managed Array Fields

### DISCUSSION THREAD

**Discussion by infirit89** [StudioCherno/Coral#12] [Q&A]
**How do I get arrays from fields?**

At first I thought something like this would do the trick:
`Coral::Array<int> arr = managedObject.GetFieldValue<Coral::Array<int>>("IntArrayField");`
but `arr` always has a length of 0.

I also looked at the examples and tests but couldn't find what I needed.

**Comment by peter1745** [comment]
That is strange, what you've written there should be how it's done, however if you're using the main branch of Coral it's possible that it's bugged, it's definitely working in the `dev` branch.
I'll probably be merging dev to main in a month at most

**Reply by infirit89** [reply]
Hey, I'm sorry I forgot to mention I was working with Coral on the `dev` branch.

I'm using the Testing project inside Coral for demonstration. This is what happens when I try to retrieve the field `IntArrayTest`:
![image](https://github.com/StudioCherno/Coral/assets/47078324/fb735fc8-35b3-426d-a289-8269a320b401)

**Reply by infirit89** [reply]
I think I might have found out what is going on.
I pinned down the problem to the method `MarshalReturnValue` in the `Marshalling` class, specifically if the type is a one dimesional array:
![image](https://github.com/StudioCherno/Coral/assets/47078324/977c4406-fc78-4a08-8a7e-87c8bde8728b)

Even though `fieldArray` is a `GCHandle` to an `Array` object, for some reason `AddrOfPinnedObject` returns a pointer to the start of the array's data. The array's length however is an 8 byte integer right before that (used visual studio's memory view to find that out). So I did this:
![image](https://github.com/StudioCherno/Coral/assets/47078324/eec45c01-d8a9-4839-ac5b-abff9a825e3e)

Just used the `ArrayContainer` struct to add the length and:
![image](https://github.com/StudioCherno/Coral/assets/47078324/d037cb70-464a-45c5-ae32-84f70a7165cf)
